### PR TITLE
feat(canvas): Canvas Connector data fetching 2/4

### PR DESCRIPTION
**Pull request by benwu408**
## Description

Add the CanvasConnector class with methods for listing courses, pages, assignments, and announcements from the Canvas API. Convert each resource type into Onyx Document objects with HTML stripping, human-readable due dates, and nullable timestamp handling. Register DocumentSource.CANVAS in the connector registry, add the CE permission stub (access.py) with EE dynamic dispatch via fetch_versioned_implementation, and implement load_credentials / validate_connector_settings with full error mapping. 

Includes unit tests for credentials, document conversion, and settings validation.

## How Has This Been Tested?

Unit tests covering credential loading (success, missing credentials, invalid token), document conversion for all 3 resource types (pages, assignments, announcements — including nullable fields and HTML stripping), and validate_connector_settings (success, expired credential, rate limiting, unexpected errors).

## PR Stack

**Stacked PRs on my fork (each targets the previous branch for isolated review):**
1. [#1 — Scaffold: API client, data models, connector registration](https://github.com/benwu408/onyx/pull/1)
2. [#2 — Connector data fetching and conversion](https://github.com/benwu408/onyx/pull/2)
3. [#3 — Checkpoint-based indexing](https://github.com/benwu408/onyx/pull/3)
4. [#4 — Permission sync and frontend registration](https://github.com/benwu408/onyx/pull/4)

**PRs on upstream onyx (all target main):**
1. [#9385 — Scaffold](https://github.com/onyx-dot-app/onyx/pull/9385)
2. [#9386 — Connector data](https://github.com/onyx-dot-app/onyx/pull/9386)
3. [#9387 — Checkpoint](https://github.com/onyx-dot-app/onyx/pull/9387)
4. [#9388 — Perm sync + frontend](https://github.com/onyx-dot-app/onyx/pull/9388)

### CODE REVIEW

File: backend/onyx/connectors/canvas/connector.py [hunk 1]
```diff
@@ -0,0 +1,421 @@
...
+                    title=p["title"],
+                    body=p.get("body"),
+                    created_at=p["created_at"],
+                    updated_at=p["updated_at"],
+                    course_id=course_id,
+                )
+                for p in response
+            )
+            if not next_url:
+                break
```
[the remark this answers is not in the document]
Comment (benwu408) [reply backend/onyx/connectors/canvas/connector.py]: not true, params are in pagination url, adding again would duplicate

File: backend/onyx/connectors/canvas/connector.py [hunk 2]
```diff
@@ -0,0 +1,413 @@
...
+        try:
+            self._canvas_client.get("courses", params={"per_page": "1"})
+        except CanvasClientRequestFailedError as e:
+            if e.status_code == 401:
+                raise ConnectorMissingCredentialError(
+                    "Invalid Canvas API token"
+                )
+            raise
+
+        return None
```
[the remark this answers is not in the document]
Comment (benwu408) [reply backend/onyx/connectors/canvas/connector.py]: load_credentials validates the token works at credential-load time. validate_connector_settings is called independently from the UI settings page. may not always be called together

File: backend/onyx/connectors/canvas/connector.py [hunk 3]
```diff
@@ -0,0 +1,414 @@
...
+    def _get_course_permissions(self, course_id: int) -> ExternalAccess | None:
+        """Get course permissions with caching."""
+        if course_id not in self._course_permissions_cache:
+            self._course_permissions_cache[course_id] = get_course_permissions(
+                canvas_client=self.canvas_client,
+                course_id=course_id,
+            )
+        return self._course_permissions_cache[course_id]
+
+    @retry(tries=3, delay=1, backoff=2)
```
[the remark this answers is not in the document]
Comment (benwu408) [reply backend/onyx/connectors/canvas/connector.py:207]: codebase wide pattern. other connectors use same format of retries regardless of error.

File: backend/onyx/connectors/canvas/connector.py [hunk 4]
```diff
@@ -0,0 +1,414 @@
...
+                    params={
+                        "per_page": "100",
+                        "enrollment_state": "active",
+                    },
+                )
+                first_request = False
+            else:
+                response, next_url = self.canvas_client.get(
+                    "", full_url=next_url
+                )
```
[the remark this answers is not in the document]
Comment (benwu408) [reply backend/onyx/connectors/canvas/connector.py]: The first_request flag pattern separates "build URL from endpoint + params" (first request) from "follow pagination link" (subsequent requests), which is the correct semantic distinction. The suggested alternative hardcodes query strings and bypasses the client's URL building, which is more fragile.

File: backend/onyx/connectors/canvas/connector.py [hunk 5]
```diff
@@ -0,0 +1,458 @@
...
+
+    Invariant:
+        If current_course_index is incremented, stage must be reset to
+        "pages" and next_url must be reset to None.
+    """
+
+    course_ids: list[int] = []
+    current_course_index: int = 0
+    stage: CanvasStage = "pages"
+    next_url: str | None = None
```
[the remark this answers is not in the document]
Comment (benwu408) [reply backend/onyx/connectors/canvas/connector.py:169]: no other checkpoint classes have a has_more default

File: backend/onyx/connectors/canvas/connector.py [hunk 6]
```diff
@@ -0,0 +1,491 @@
...
+            )
+
+    @override
+    def load_from_checkpoint(
+        self,
+        start: SecondsSinceUnixEpoch,
+        end: SecondsSinceUnixEpoch,
+        checkpoint: CanvasConnectorCheckpoint,
+    ) -> CheckpointOutput[CanvasConnectorCheckpoint]:
+        raise NotImplementedError("Implemented in follow-up PR")
```
[the remark this answers is not in the document]
Comment (benwu408) [reply backend/onyx/connectors/canvas/connector.py]: kept as structure for following PR

File: backend/onyx/connectors/canvas/connector.py [hunk 7]
```diff
@@ -0,0 +1,524 @@
...
+
+    Invariant:
+        If current_course_index is incremented, stage must be reset to
+        "pages" and next_url must be reset to None.
+    """
+
+    course_ids: list[int] = []
+    current_course_index: int = 0
+    stage: CanvasStage = "pages"
+    next_url: str | None = None
```
[the remark this answers is not in the document]
Comment (benwu408) [reply backend/onyx/connectors/canvas/connector.py:169]: not true, that notion has_more is for another class. no other classes have has_more default

File: backend/onyx/connectors/canvas/connector.py [hunk 8]
```diff
@@ -0,0 +1,524 @@
...
+        courses: list[CanvasCourse] = []
+        next_url: str | None = None
+        first_request = True
+        while True:
+            if first_request:
+                response, next_url = self.canvas_client.get(
+                    "courses",
+                    params={
+                        "per_page": "100",
+                        "state[]": "available",
```
[the remark this answers is not in the document]
Comment (benwu408) [reply backend/onyx/connectors/canvas/connector.py]: only want to get info from current courses. search would be filled with meaningless data from old irrelevant courses if state wasn't limited to just available.

File: backend/onyx/connectors/canvas/connector.py [hunk 9]
```diff
@@ -0,0 +1,527 @@
...
+
+    @classmethod
+    def from_api(cls, payload: dict[str, Any]) -> "CanvasCourse":
+        return cls(
+            id=payload["id"],
+            name=payload.get("name", ""),
+            course_code=payload.get("course_code", ""),
+            created_at=payload.get("created_at", ""),
+            workflow_state=payload.get("workflow_state", ""),
+        )
```
[the remark this answers is not in the document]
Comment (benwu408) [reply backend/onyx/connectors/canvas/connector.py:79]: Primary identifiers (id, page_id) and required fields (name, html_url, url, title) are guaranteed by the Canvas API for well-formed responses. Using .get() with sentinels would add complexity for a scenario that indicates a
 fundamentally broken API response — which should fail loudly, not silently produce garbage documents with id=0 or name="".

File: backend/onyx/connectors/canvas/connector.py [hunk 10]
```diff
@@ -0,0 +1,527 @@
...
+            self._canvas_client = CanvasApiClient(
+                bearer_token=access_token,
+                canvas_base_url=self.canvas_base_url,
+            )
+            self._canvas_client.get("courses", params={"per_page": "1"})
+        except ValueError as e:
+            raise ConnectorValidationError(
+                f"Invalid Canvas base URL: {e}"
+            )
+        except OnyxError as e:
```
[the remark this answers is not in the document]
Comment (benwu408) [reply backend/onyx/connectors/canvas/connector.py:397]: Not fixing — invalid concern. ConnectorMissingCredentialError inherits from PermissionError, not OnyxError. The except OnyxError block in load_credentials will never catch it. No guard is needed

File: backend/onyx/connectors/canvas/connector.py [hunk 11]
```diff
@@ -0,0 +1,424 @@
...
+
+class CanvasAssignment(BaseModel):
+    id: int
+    name: str
+    description: str | None = None
+    html_url: str
+    course_id: int
+    created_at: str
+    updated_at: str
+    due_at: str | None = None
```
[the remark this answers is not in the document]
Comment (benwu408) [reply backend/onyx/connectors/canvas/connector.py:112]: isnt used directly, all other connectors don't use either but keep for possible future features or debugging.

File: backend/onyx/connectors/canvas/connector.py [hunk 12]
```diff
@@ -0,0 +1,444 @@
...
+                )
+
+            if not response:
+                break
+            pages.extend(
+                CanvasPage.from_api(p, course_id=course_id)
+                for p in response
+            )
+            if not next_url:
+                break
```
[the remark this answers is not in the document]
Comment (benwu408) [reply backend/onyx/connectors/canvas/connector.py]: next url link in header alr includes parameters

File: backend/onyx/connectors/canvas/connector.py [hunk 13]
```diff
@@ -0,0 +1,421 @@
...
+                    html_url=a["html_url"],
+                    posted_at=a.get("posted_at"),
+                    course_id=course_id,
+                )
+                for a in response
+            )
+            if not next_url:
+                break
+
+        return announcements
```
[the remark this answers is not in the document]
Comment (benwu408) [reply backend/onyx/connectors/canvas/connector.py:264]: wrong, canvas does return params in next link

File: backend/onyx/connectors/canvas/connector.py [hunk 14]
```diff
@@ -0,0 +1,527 @@
...
+            if e.status_code == 401:
+                raise CredentialExpiredError(
+                    "Canvas API token is invalid or expired (HTTP 401)."
+                )
+            elif e.status_code == 403:
+                raise InsufficientPermissionsError(
+                    "Canvas API token does not have sufficient permissions (HTTP 403)."
+                )
+            elif e.status_code == 429:
+                raise ConnectorValidationError(
```
[the remark this answers is not in the document]
Comment (benwu408) [reply backend/onyx/connectors/canvas/connector.py]: no one else does this. unnecessary

File: backend/onyx/connectors/canvas/connector.py [hunk 15]
```diff
@@ -13,6 +45,16 @@ class CanvasCourse(BaseModel):
     created_at: str
     workflow_state: str
 
+    @classmethod
+    def from_api(cls, payload: dict[str, Any]) -> "CanvasCourse":
+        return cls(
+            id=payload["id"],
```
Comment (justin-tahara) [inline backend/onyx/connectors/canvas/connector.py:74]: Another place where we are implicitly accepting the fact that ID exists.

File: backend/onyx/connectors/canvas/connector.py [hunk 16]
```diff
@@ -23,6 +65,20 @@ class CanvasPage(BaseModel):
...
 
+    @classmethod
+    def from_api(
+        cls, payload: dict[str, Any], course_id: int
+    ) -> "CanvasPage":
+        return cls(
+            page_id=payload["page_id"],
+            url=payload["url"],
+            title=payload["title"],
+            body=payload.get("body"),
```
Comment (justin-tahara) [inline backend/onyx/connectors/canvas/connector.py:97]: Should have a proper default value like the other ones.

File: backend/onyx/connectors/canvas/connector.py [hunk 17]
```diff
@@ -34,6 +90,21 @@ class CanvasAssignment(BaseModel):
...
     due_at: str | None = None
 
+    @classmethod
+    def from_api(
+        cls, payload: dict[str, Any], course_id: int
+    ) -> "CanvasAssignment":
+        return cls(
+            id=payload["id"],
+            name=payload["name"],
+            description=payload.get("description"),
```
Comment (justin-tahara) [inline backend/onyx/connectors/canvas/connector.py:119]: Another instance where the class being constructed is not uniform but more importantly should we have a default?

File: backend/onyx/connectors/canvas/connector.py [hunk 18]
```diff
@@ -72,3 +156,377 @@ def advance_course(self) -> None:
...
+class CanvasConnector(
+    CheckpointedConnectorWithPermSync[CanvasConnectorCheckpoint],
+    SlimConnectorWithPermSync,
+):
+    def __init__(
+        self,
+        canvas_base_url: str,
+        batch_size: int = INDEX_BATCH_SIZE,
+    ) -> None:
+        self.canvas_base_url = canvas_base_url.rstrip("/").removesuffix("/api/v1")
```
Comment (justin-tahara) [inline backend/onyx/connectors/canvas/connector.py:187]: Any reason this is being done here rather than when we setup the Canvas Connector? Should we sanitize the base url before passing it into this class initialization?

File: backend/onyx/connectors/canvas/connector.py [hunk 19]
```diff
@@ -72,3 +156,377 @@ def advance_course(self) -> None:
...
+    @property
+    def canvas_client(self) -> CanvasApiClient:
+        if self._canvas_client is None:
+            raise ConnectorMissingCredentialError("Canvas")
+        return self._canvas_client
+
+    def _get_course_permissions(self, course_id: int) -> ExternalAccess | None:
+        """Get course permissions with caching."""
+        if course_id not in self._course_permissions_cache:
+            self._course_permissions_cache[course_id] = get_course_permissions(
```
Comment (justin-tahara) [inline backend/onyx/connectors/canvas/connector.py:201]: Should we be checking if this is properly returning something before we set it within the ExternalAccess

Comment (justin-tahara) [reply backend/onyx/connectors/canvas/connector.py:201]: I don't think there is any exception handling in the function you are calling either

File: backend/onyx/connectors/canvas/connector.py [hunk 20]
```diff
@@ -72,3 +156,377 @@ def advance_course(self) -> None:
...
+
+    @retry(tries=3, delay=1, backoff=2)
+    def _list_courses(self) -> list[CanvasCourse]:
+        """Fetch all courses accessible to the authenticated user."""
+        logger.debug("Fetching Canvas courses")
+
+        courses: list[CanvasCourse] = []
+        next_url: str | None = None
+        first_request = True
+        while True:
```
Comment (justin-tahara) [inline backend/onyx/connectors/canvas/connector.py]: Is there a way to not have an infinite loop that we exit from?

File: backend/onyx/connectors/canvas/connector.py [hunk 21]
```diff
@@ -72,3 +156,377 @@ def advance_course(self) -> None:
...
+            if first_request:
```
Comment (justin-tahara) [inline backend/onyx/connectors/canvas/connector.py]: Just take this out and then have a loop through the next pages.

File: backend/onyx/connectors/canvas/connector.py [hunk 22]
```diff
@@ -72,3 +156,377 @@ def advance_course(self) -> None:
...
+                response, next_url = self.canvas_client.get(
+                    f"courses/{course_id}/pages",
+                    params={
+                        "per_page": "100",
+                        "include[]": "body",
+                        "published": "true",
+                    },
+                )
+                first_request = False
```
Comment (justin-tahara) [inline backend/onyx/connectors/canvas/connector.py]: Same pattern here. Probably worth taking this out and then just looping on page 2 and beyond

File: backend/onyx/connectors/canvas/connector.py [hunk 23]
```diff
@@ -72,3 +156,377 @@ def advance_course(self) -> None:
...
+            else:
+                assert next_url is not None
+                response, next_url = self.canvas_client.get(
+                    "", full_url=next_url
+                )
+
+            if not response:
+                break
+            courses.extend(
+                CanvasCourse.from_api(course)
```
Comment (justin-tahara) [inline backend/onyx/connectors/canvas/connector.py]: Make a giant helper that does the following: 

Create the first endpoint and starts that process and then the second part of the helper function takes in a different arg like CanvasCourse, CanvasPage and other ones to get the responses for that Canvas resource. That way you can reduce a ton of the code in this file. It should simplify the tests as well.

File: backend/onyx/connectors/canvas/connector.py [hunk 24]
```diff
@@ -72,3 +156,377 @@ def advance_course(self) -> None:
...
+            id=f"canvas-page-{page.course_id}-{page.page_id}",
+            sections=cast(list[TextSection | ImageSection], sections),
+            source=DocumentSource.CANVAS,
+            semantic_identifier=page.title or f"Page {page.page_id}",
+            doc_updated_at=datetime.fromisoformat(
+                page.updated_at.replace("Z", "+00:00")
+            ).astimezone(timezone.utc)
+            if page.updated_at
+            else None,
+            metadata={"course_id": str(page.course_id), "type": "page"},
```
Comment (justin-tahara) [inline backend/onyx/connectors/canvas/connector.py]: Let's make this cleaner

Comment (justin-tahara) [reply backend/onyx/connectors/canvas/connector.py]: Like we can get these values before the return statement to make this more readable

File: backend/onyx/connectors/canvas/connector.py [hunk 25]
```diff
@@ -72,3 +156,377 @@ def advance_course(self) -> None:
...
+
+        return Document(
+            id=f"canvas-announcement-{announcement.course_id}-{announcement.id}",
+            sections=cast(list[TextSection | ImageSection], sections),
+            source=DocumentSource.CANVAS,
+            semantic_identifier=announcement.title
+            or f"Announcement {announcement.id}",
+            doc_updated_at=doc_updated_at,
+            metadata={"course_id": str(announcement.course_id), "type": "announcement"},
+        )
```
Comment (justin-tahara) [inline backend/onyx/connectors/canvas/connector.py]: I think the Document creation can be made into a helper function as well

File: backend/onyx/connectors/canvas/connector.py [hunk 26]
```diff
@@ -72,3 +156,377 @@ def advance_course(self) -> None:
...
+        access_token = credentials.get("canvas_access_token")
+        if not access_token:
+            raise ConnectorMissingCredentialError("Canvas")
+
+        try:
+            client = CanvasApiClient(
+                bearer_token=access_token,
+                canvas_base_url=self.canvas_base_url,
+            )
+            client.get("courses", params={"per_page": "1"})
```
Comment (justin-tahara) [inline backend/onyx/connectors/canvas/connector.py:394]: Is this your attempt to check if the the permission works? What if there are no courses? is that a possibility?

File: backend/onyx/connectors/canvas/connector.py [hunk 27]
```diff
@@ -72,3 +156,377 @@ def advance_course(self) -> None:
...
+                )
+
+        self._canvas_client = client
+        return None
+
+    @override
+    def validate_connector_settings(self) -> None:
+        """Validate Canvas connector settings by testing API access."""
+        try:
+            self.canvas_client.get("courses", params={"per_page": "1"})
```
Comment (justin-tahara) [inline backend/onyx/connectors/canvas/connector.py:407]: You have the same check as the other one. Is this enough? Is this equivalent to sending a chat message to the OpenAI chat endpoint?

File: backend/onyx/connectors/canvas/connector.py [hunk 28]
```diff
@@ -72,3 +156,377 @@ def advance_course(self) -> None:
...
+                )
+            elif e.status_code == 429:
+                raise ConnectorValidationError(
+                    "Validation failed due to Canvas rate-limits being exceeded (HTTP 429). "
+                    "Please try again later."
+                )
+            else:
+                raise UnexpectedValidationError(
+                    f"Unexpected Canvas HTTP error (status={e.status_code}): {e}"
+                )
```
Comment (justin-tahara) [inline backend/onyx/connectors/canvas/connector.py]: Let's simplify this. I think it should be put in a helper function

File: backend/tests/unit/onyx/connectors/canvas/test_canvas_connector.py [hunk 29]
```diff
@@ -379,3 +460,460 @@ def test_rejects_non_https_link(self) -> None:
...
+        self, mock_requests: MagicMock
+    ) -> None:
+        success_resp = _mock_response(json_data=[_mock_course()])
+        fail_resp = _mock_response(403, {})
+        mock_requests.get.side_effect = [success_resp, fail_resp]
+        connector = CanvasConnector(canvas_base_url=FAKE_BASE_URL)
+        connector.load_credentials({"canvas_access_token": FAKE_TOKEN})
+
+        with pytest.raises(InsufficientPermissionsError):
+            connector.validate_connector_settings()
```
Comment (justin-tahara) [inline backend/tests/unit/onyx/connectors/canvas/test_canvas_connector.py]: I think you should be able to make a helper function where you pass in the error code and the error that should be raised and you can reduce the lines here

File: backend/onyx/connectors/canvas/connector.py (hunk 15 above)
Comment (benwu408) [reply backend/onyx/connectors/canvas/connector.py:74]: Ok so for the from api methods, the ones that use [] to get it are required fields, and then the ones that use get() are optional. but I guess when I was making the classes I made some optional and some not, so the ones that are just get() with no default are ones that the class sets up as str | none, and the rest the default gets set to "" cause the class can't handle none. but that was really dumb so I made all the optional fields str | none and just used get() so that if doesn't exist it'll just return none and the class can handle that now. the ones that are [] are required, so we'll get errors if for example we try to get data from a document with no title which should never happen.

File: backend/onyx/connectors/canvas/connector.py (hunk 16 above)
Comment (benwu408) [reply backend/onyx/connectors/canvas/connector.py:97]: see comment on line 51

File: backend/onyx/connectors/canvas/connector.py (hunk 17 above)
Comment (benwu408) [reply backend/onyx/connectors/canvas/connector.py:119]: see reply on line 51

File: backend/onyx/connectors/canvas/connector.py (hunk 18 above)
Comment (benwu408) [reply backend/onyx/connectors/canvas/connector.py:187]: All connector classes get initialized with the values that are provided by admins when setting up a connector in the frontend. this is just normalizing the url given by frontend. No step in-between passing url into class init.

File: backend/onyx/connectors/canvas/connector.py (hunk 19 above)
Comment (benwu408) [reply backend/onyx/connectors/canvas/connector.py:201]: For first question if ur asking if it's returning the right things, in get course permissions it returns none if not enterprise edition, so thats why none is an option. if it is EE, we build the external access, which validates while its built since its a pydantic model. 

for the second question, get course permissions is in PR2 because it is part of a perm synced connector, but all it really does is call get course permissions, which is actually another wrapper that checks if EE or CE, and then if EE it uses the real perm sync logic that is built in PR4. that does include exception handling and returns an empty external access on failure and logs errors.

File: backend/onyx/connectors/canvas/connector.py (hunk 21 above)
Comment (benwu408) [reply backend/onyx/connectors/canvas/connector.py]: added a paginate function in client.py that takes in permissions, endpoint, and then returns it paginated for all the list ___ functions and the get course permission function!

File: backend/onyx/connectors/canvas/connector.py (hunk 23 above)
Comment (benwu408) [reply backend/onyx/connectors/canvas/connector.py]: see reply on 199

File: backend/onyx/connectors/canvas/connector.py (hunk 22 above)
Comment (benwu408) [reply backend/onyx/connectors/canvas/connector.py]: see comment on 199

File: backend/onyx/connectors/canvas/connector.py (hunk 26 above)
Comment (benwu408) [reply backend/onyx/connectors/canvas/connector.py:394]: checks if link, token are correct and canvas is reachable. if no courses it returns empty list, still 200 response.

File: backend/onyx/connectors/canvas/connector.py (hunk 27 above)
Comment (benwu408) [reply backend/onyx/connectors/canvas/connector.py:407]: load credentials is for setting up the client when canvas connector is initialized. validate connector settings checks again later, in situations like if a token was revoked or something. verifies canvas is reachable with link and token. more so equivalent to openai /models endpoint that lists models. lightest possible call, will always succeed if link and token are valid, no weird edge cases.

File: backend/onyx/connectors/canvas/connector.py (hunk 28 above)
Comment (benwu408) [reply backend/onyx/connectors/canvas/connector.py]: done

File: backend/tests/unit/onyx/connectors/canvas/test_canvas_connector.py (hunk 29 above)
Comment (benwu408) [reply backend/tests/unit/onyx/connectors/canvas/test_canvas_connector.py]: done

**Review by justin-tahara — APPROVED** [review]
